From c5341b5c4de36b781484dab6f6c345f41f42ba75 Mon Sep 17 00:00:00 2001
From: Padmarao Begari <padmarao.begari@amd.com>
Date: Fri, 12 Sep 2025 15:35:38 +0530
Subject: [PATCH] FWU: Add platform hook support for fwu metata

FWU metadata information is unavailable for platform-specific
operations since FWU initialization has not yet occurred.
The initialization function is invoked as part of the main
loop event. To address this, the FWU platform hook function
is introduced during FWU initialization, allowing metadata
processing with platform-specific operations.

Signed-off-by: Padmarao Begari <padmarao.begari@amd.com>
Link: https://lore.kernel.org/r/20250912100539.4127378-2-padmarao.begari@amd.com
Signed-off-by: Michal Simek <michal.simek@amd.com>
Signed-off-by: Dario Binacchi <dario.binacchi@amarulasolutions.com>
Upstream-Status: Backport [https://source.denx.de/u-boot/u-boot/-/commit/aa4e7f70837fc7f714822de234d30fb13fb29d14]
---
 include/fwu.h         | 11 +++++++++++
 lib/fwu_updates/fwu.c | 18 ++++++++++++++++++
 2 files changed, 29 insertions(+)

diff --git a/include/fwu.h b/include/fwu.h
index 77ec65e61807..865c239b66a4 100644
--- a/include/fwu.h
+++ b/include/fwu.h
@@ -129,6 +129,17 @@ int fwu_read_mdata(struct udevice *dev, struct fwu_mdata *mdata,
 int fwu_write_mdata(struct udevice *dev, struct fwu_mdata *mdata,
 		    bool primary, uint32_t size);
 
+/**
+ * fwu_platform_hook() - Platform specific processing with FWU metadata
+ * @dev: FWU metadata device
+ * @data: FWU metadata
+ *
+ * Provide a platform specific function for processing with the FWU metadata.
+ *
+ * Return: 0 if OK, -ve on error
+ */
+int fwu_platform_hook(struct udevice *dev, struct fwu_data *data);
+
 /**
  * fwu_get_mdata() - Read, verify and return the FWU metadata
  *
diff --git a/lib/fwu_updates/fwu.c b/lib/fwu_updates/fwu.c
index f5e0315b796f..50793a7784c9 100644
--- a/lib/fwu_updates/fwu.c
+++ b/lib/fwu_updates/fwu.c
@@ -629,6 +629,16 @@ __weak void fwu_plat_get_bootidx(uint *boot_idx)
 		*boot_idx = 0; /* Dummy value */
 }
 
+/**
+ * fwu_platform_hook() - Platform specific processing with FWU metadata
+ *
+ * Return: 0 if OK, -ve on error
+ */
+__weak int fwu_platform_hook(struct udevice *dev, struct fwu_data *data)
+{
+	return 0;
+}
+
 /**
  * fwu_update_checks_pass() - Check if FWU update can be done
  *
@@ -686,6 +696,7 @@ static int fwu_boottime_checks(void *ctx, struct event *event)
 {
 	int ret;
 	u32 boot_idx, active_idx;
+	struct fwu_data *data;
 
 	ret = uclass_first_device_err(UCLASS_FWU_MDATA, &g_dev);
 	if (ret) {
@@ -744,6 +755,13 @@ static int fwu_boottime_checks(void *ctx, struct event *event)
 	if (!ret)
 		boottime_check = 1;
 
+	data = fwu_get_data();
+	ret = fwu_platform_hook(g_dev, data);
+	if (ret) {
+		log_err("fwu_platform_hook() failed\n");
+		return ret;
+	}
+
 	return 0;
 }
 EVENT_SPY(EVT_MAIN_LOOP, fwu_boottime_checks);
-- 
2.43.0

